sdk/go/common/workspace/plugins.go

```diff
@@ -199,6 +199,101 @@ func (source *getPulumiSource) Download(
 	return getHTTPResponse(req)
 }
 
+// gitlabSource can download a plugin from gitlab releases
+type gitlabSource struct {
+	host    string
+	project string
+	name    string
+	kind    PluginKind
+
+	token string
+}
+
+// Creates a new github source adding authentication data in the environment, if it exists
+func newGitlabSource(url *url.URL, name string, kind PluginKind) (*gitlabSource, error) {
+	contract.Assert(url.Scheme == "gitlab")
+
+	host := url.Host
+	parts := strings.Split(strings.Trim(url.Path, "/"), "/")
+
+	if host == "" {
+		return nil, fmt.Errorf("gitlab:// url must have a host part, was: %s", url.String())
+	}
+
+	if len(parts) != 1 {
+		return nil, fmt.Errorf(
+			"gitlab:// url must have the format <host>/<project>, was: %s",
+			url.String())
+	}
+
+	project := parts[0]
+
+	return &gitlabSource{
+		host:    host,
+		project: project,
+		name:    name,
+		kind:    kind,
+
+		token: os.Getenv("GITLAB_TOKEN"),
+	}, nil
+}
+
+func (source *gitlabSource) HasAuthentication() bool {
+	return source.token != ""
+}
+
+func (source *gitlabSource) GetLatestVersion(
+	getHTTPResponse func(*http.Request) (io.ReadCloser, int64, error)) (*semver.Version, error) {
+	releaseURL := fmt.Sprintf(
+		"https://%s/api/v4/projects/%s/releases/permalink/latest",
+		source.host, source.project)
+	logging.V(9).Infof("plugin GitLab releases url: %s", releaseURL)
+	req, err := buildHTTPRequest(releaseURL, fmt.Sprintf("Bearer %s", source.token))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", "application/json")
+	resp, length, err := getHTTPResponse(req)
+	if err != nil {
+		return nil, err
+	}
+	jsonBody, err := io.ReadAll(resp)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshal gitlab response len(%d): %s", length, err.Error())
+	}
+	release := struct {
+		TagName string `json:"tag_name"`
+	}{}
+	err = json.Unmarshal(jsonBody, &release)
+	if err != nil {
+		return nil, err
+	}
+	parsedVersion, err := semver.ParseTolerant(release.TagName)
+	if err != nil {
+		return nil, fmt.Errorf("invalid plugin semver: %w", err)
```

Comment:
Seeing as the tag name comes from The Internet, we might want to include it in the error message.

```suggestion
		return nil, fmt.Errorf("invalid plugin tag %q: %w", release.TagName, err)
```